src/core/serializer/portable_network_graphics_serializer.cpp

```diff
@@ -1,31 +1,320 @@
+	int chunkLengthInt = (int)imageData.size();
+	append(chunkLength, intToBytes(chunkLengthInt));
+
+	std::vector<unsigned char> chunkType{ 73, 68, 65, 84 };
+
+	std::vector<unsigned char> chunk;
+
+	// Chunk Header
+	append(chunk, chunkLength);
+	append(chunk, chunkType);
+
+	// Image Data
+	append(chunk, imageData);
+
+	// CRC
+	append(chunk, buildCRC(chunk, chunkLengthInt));
+
+	return chunk;
+}
 
-	std::stringstream ss;
-	ss << signature;
+std::vector<unsigned char> PortableNetworkGraphicsSerializer::buildIENDChunk()
+{
+	std::vector<unsigned char> chunkLength;
+	int chunkLengthInt = 0;
+	append(chunkLength, intToBytes(chunkLengthInt));
+
+	std::vector<unsigned char> chunkType{ 73, 69, 78, 68 };
+
+	std::vector<unsigned char> chunk;
+
+	// Chunk Header
+	append(chunk, chunkLength);
+	append(chunk, chunkType);
+
+	// CRC
+	append(chunk, buildCRC(chunk, chunkLengthInt));
+
+	return chunk;
+}
+
+unsigned char setWithFCHECK(unsigned char cmf, unsigned char flg)
+{
+	// Data size is assumed to be 16 bits
+
+	// TODO: Figure out algebraic formula for determining the FCHECK, without doing a for loop
+
+	std::bitset<8> cmfBits = cmf;
+	std::bitset<8> flgBits = flg;
+
+	for (unsigned int i = 0b00000; i <= 0b11111; ++i)
+	{
+		std::bitset<16> checkBits;
+		std::bitset<5> indexBits = i;
+
+		// Set the iterated number
+		checkBits[0] = indexBits[0];
+		checkBits[1] = indexBits[1];
+		checkBits[2] = indexBits[2];
+		checkBits[3] = indexBits[3];
+		checkBits[4] = indexBits[4];
+
+		// Set the flg
+		checkBits[5] = flgBits[5]; // FDICT
+		checkBits[6] = flgBits[6]; // FLEVEL
+		checkBits[7] = flgBits[7]; // FLEVEL
+
+		// Set the cmf
+		checkBits[8] = cmfBits[0];
+		checkBits[9] = cmfBits[1];
+		checkBits[10] = cmfBits[2];
+		checkBits[11] = cmfBits[3];
+		checkBits[12] = cmfBits[4];
+		checkBits[13] = cmfBits[5];
+		checkBits[14] = cmfBits[6];
+		checkBits[15] = cmfBits[7];
+
+		auto check = static_cast<unsigned int>(checkBits.to_ulong());
+
+		if (check % 31 == 0)
+		{
+			std::bitset<8> resultBits;
+
+			resultBits[0] = checkBits[0];
+			resultBits[1] = checkBits[1];
+			resultBits[2] = checkBits[2];
+			resultBits[3] = checkBits[3];
+			resultBits[4] = checkBits[4];
+			resultBits[5] = checkBits[5];
+			resultBits[6] = checkBits[6];
+			resultBits[7] = checkBits[7];
+
+			return static_cast<unsigned char>(resultBits.to_ulong());
+		}
+	}
+
+	throw std::logic_error("Could not compute a valid FCHECK when encoding PNG file");
+}
+
+bool inline getBit(CppRayTracerChallenge::Core::Compression::DeflateBlock::DeflateBitset const& bitset, unsigned int index)
+{
+	if (index >= bitset.size())
+		return false;
 
-	return ss.str();
+	return bitset.test(index);
+}
+
+unsigned char convertColorValue(float colorValue) // TODO: Merge with PPM to base class or helper class
+{
+	float scaledColor = colorValue * 255;
+	scaledColor = std::ceil(scaledColor);
+	scaledColor = std::clamp(scaledColor, 0.0f, 255.0f);
+
+	return static_cast<unsigned char>(scaledColor);
+}
```

Comment:
Move to the base class, or a conversion class

Reply:
Nevermind, each serializer should have its own implementation because the bit depth can differ